File: bayan/bayan/complete_vocabulary.py
# -*- coding: utf-8 -*-
"""
مدير قاموس الكلمات الأساسية الكامل
Complete Foundation Vocabulary Manager

المصدر الأصلي: TypeScript في /vocabulary/completeVocabulary.ts
Original source: TypeScript in /vocabulary/completeVocabulary.ts
"""

from typing import List, Dict
from .foundation_vocabulary import (
    FoundationVocabulary,
    FoundationWord,
    FoundationWordType,
    FoundationCategory
)
from .foundation_vocabulary_extended import add_extended_vocabulary
from .foundation_vocabulary_complete import add_complete_vocabulary


class CompleteFoundationVocabulary:
    """
    مدير قاموس الكلمات الأساسية الكامل
    Complete Foundation Vocabulary Manager
    """
    
    def __init__(self):
        self.vocab = FoundationVocabulary()
        self._load_all_vocabulary()
    
    def _load_all_vocabulary(self) -> None:
        """
        تحميل جميع الكلمات الأساسية
        Load all foundation words
        """
        # الكلمات الأساسية موجودة في FoundationVocabulary
        # Foundation words are already in FoundationVocabulary
        
        # تحميل الكلمات الموسعة
        # Load extended words
        add_extended_vocabulary(self.vocab)
        
        # تحميل الكلمات الكاملة
        # Load complete words
        add_complete_vocabulary(self.vocab)
    
    def get_vocabulary(self) -> FoundationVocabulary:
        """الحصول على القاموس / Get vocabulary"""
        return self.vocab
    
    def get_word(self, arabic: str) -> FoundationWord:
        """الحصول على كلمة / Get word"""
        return self.vocab.get_word(arabic)
    
    def get_words_by_category(self, category: FoundationCategory) -> List[FoundationWord]:
        """الحصول على كلمات حسب الفئة / Get words by category"""
        return self.vocab.get_words_by_category(category)
    
    def get_words_by_type(self, word_type: FoundationWordType) -> List[FoundationWord]:
        """الحصول على كلمات حسب النوع / Get words by type"""
        return self.vocab.get_words_by_type(word_type)
    
    def find_related_words(self, arabic: str) -> List[FoundationWord]:
        """إيجاد كلمات مرتبطة / Find related words"""
        return self.vocab.find_related_words(arabic)
    
    def get_all_words(self) -> List[FoundationWord]:
        """الحصول على جميع الكلمات / Get all words"""
        return self.vocab.get_all_words()
    
    def get_word_count(self) -> int:
        """عدد الكلمات / Word count"""
        return self.vocab.get_word_count()
    
    def get_statistics(self) -> Dict:
        """
        الحصول على إحصائيات
        Get statistics
        """
        all_words = self.vocab.get_all_words()
        by_type = {}
        by_category = {}
        
        for word in all_words:
            # Count by type
            type_key = word.word_type
            by_type[type_key] = by_type.get(type_key, 0) + 1
            
            # Count by category
            category_key = word.category
            by_category[category_key] = by_category.get(category_key, 0) + 1
        
        return {
            'total_words': len(all_words),
            'by_type': by_type,
            'by_category': by_category
        }
    
    def print_statistics(self) -> None:
        """طباعة الإحصائيات / Print statistics"""
        stats = self.get_statistics()
        
        print('\n📊 إحصائيات القاموس الأساسي / Foundation Vocabulary Statistics')
        print('=' * 60)
        print(f'\n📝 إجمالي الكلمات / Total Words: {stats["total_words"]}')
        
        print('\n📂 حسب النوع / By Type:')
        print('─' * 60)
        for word_type, count in stats['by_type'].items():
            print(f'   {word_type.value}: {count}')
        
        print('\n🏷️  حسب الفئة / By Category:')
        print('─' * 60)
        for category, count in stats['by_category'].items():
            print(f'   {category.value}: {count}')
        print('=' * 60)
    
    def show_examples(self) -> None:
        """عرض أمثلة / Show examples"""
        categories = [
            FoundationCategory.INITIAL_ENVIRONMENT,
            FoundationCategory.ENTITY_EXISTENCE,
            FoundationCategory.PHYSICAL,
            FoundationCategory.BASIC_ACTIONS
        ]
        
        print('\n📚 أمثلة من القاموس / Examples from Vocabulary')
        print('=' * 60)
        
        for category in categories:
            words = self.vocab.get_words_by_category(category)
            if words:
                print(f'\n🏷️  {category.value}:')
                example_words = words[:3]
                for word in example_words:
                    english = word.english or 'N/A'
                    print(f'   • {word.arabic} ({english}): {word.core_meaning}')
        print('=' * 60)
    
    def search_by_meaning(self, search_term: str) -> List[FoundationWord]:
        """
        البحث بالمعنى
        Search by meaning
        """
        all_words = self.vocab.get_all_words()
        return [
            word for word in all_words
            if search_term in word.core_meaning
            or search_term in word.arabic
            or (word.english and search_term in word.english)
        ]
    
    def get_related_words_tree(self, arabic: str, depth: int = 2) -> Dict[str, List[FoundationWord]]:
        """
        الحصول على شجرة الكلمات المرتبطة
        Get related words tree
        """
        tree = {}
        visited = set()
        
        def explore(word: str, current_depth: int):
            if current_depth > depth or word in visited:
                return
            
            visited.add(word)
            related = self.vocab.find_related_words(word)
            tree[word] = related
            
            if current_depth < depth:
                for related_word in related:
                    explore(related_word.arabic, current_depth + 1)
        
        explore(arabic, 0)
        return tree


# Singleton instance
_vocabulary_instance: CompleteFoundationVocabulary = None


def get_complete_vocabulary() -> CompleteFoundationVocabulary:
    """
    الحصول على نسخة واحدة من القاموس
    Get singleton vocabulary instance
    """
    global _vocabulary_instance
    if _vocabulary_instance is None:
        _vocabulary_instance = CompleteFoundationVocabulary()
    return _vocabulary_instance


def reset_vocabulary() -> None:
    """
    إعادة تعيين القاموس
    Reset vocabulary
    """
    global _vocabulary_instance
    _vocabulary_instance = None

File: vocabulary/completeVocabulary.ts
/**
 * مدير قاموس الكلمات الأساسية الكامل
 * Complete Foundation Vocabulary Manager
 * 
 * @author Basel Yahya Abdullah
 */

import { 
  FoundationVocabulary, 
  FoundationWord, 
  FoundationWordType, 
  FoundationCategory 
} from './foundationVocabulary';
import { addExtendedVocabulary } from './foundationVocabularyExtended';
import { addCompleteVocabulary } from './foundationVocabularyComplete';

/**
 * مدير قاموس الكلمات الأساسية الكامل
 * Complete Foundation Vocabulary Manager
 */
export class CompleteFoundationVocabulary {
  private vocab: FoundationVocabulary;

  constructor() {
    this.vocab = new FoundationVocabulary();
    this.loadAllVocabulary();
  }

  /**
   * تحميل جميع الكلمات الأساسية
   * Load all foundation words
   */
  private loadAllVocabulary(): void {
    // الكلمات الأساسية موجودة في FoundationVocabulary
    // Foundation words are already in FoundationVocabulary
    
    // تحميل الكلمات الموسعة
    // Load extended words
    addExtendedVocabulary(this.vocab);
    
    // تحميل الكلمات الكاملة
    // Load complete words
    addCompleteVocabulary(this.vocab);
  }

  /**
   * الحصول على القاموس
   * Get vocabulary
   */
  getVocabulary(): FoundationVocabulary {
    return this.vocab;
  }

  /**
   * الحصول على كلمة
   * Get word
   */
  getWord(arabic: string): FoundationWord | undefined {
    return this.vocab.getWord(arabic);
  }

  /**
   * الحصول على كلمات حسب الفئة
   * Get words by category
   */
  getWordsByCategory(category: FoundationCategory): FoundationWord[] {
    return this.vocab.getWordsByCategory(category);
  }

  /**
   * الحصول على كلمات حسب النوع
   * Get words by type
   */
  getWordsByType(type: FoundationWordType): FoundationWord[] {
    return this.vocab.getWordsByType(type);
  }

  /**
   * إيجاد كلمات مرتبطة
   * Find related words
   */
  findRelatedWords(arabic: string): FoundationWord[] {
    return this.vocab.findRelatedWords(arabic);
  }

  /**
   * الحصول على جميع الكلمات
   * Get all words
   */
  getAllWords(): FoundationWord[] {
    return this.vocab.getAllWords();
  }

  /**
   * عدد الكلمات
   * Word count
   */
  getWordCount(): number {
    return this.vocab.getWordCount();
  }

  /**
   * الحصول على إحصائيات
   * Get statistics
   */
  getStatistics(): {
    totalWords: number;
    byType: Map<FoundationWordType, number>;
    byCategory: Map<FoundationCategory, number>;
  } {
    const allWords = this.vocab.getAllWords();
    const byType = new Map<FoundationWordType, number>();
    const byCategory = new Map<FoundationCategory, number>();

    for (const word of allWords) {
      // Count by type
      const typeCount = byType.get(word.type) || 0;
      byType.set(word.type, typeCount + 1);

      // Count by category
      const categoryCount = byCategory.get(word.category) || 0;
      byCategory.set(word.category, categoryCount + 1);
    }

    return {
      totalWords: allWords.length,
      byType,
      byCategory
    };
  }

  /**
   * طباعة الإحصائيات
   * Print statistics
   */
  printStatistics(): void {
    const stats = this.getStatistics();
    
    console.log('\n📊 إحصائيات القاموس الأساسي / Foundation Vocabulary Statistics');
    console.log('═'.repeat(60));
    console.log(`\n📝 إجمالي الكلمات / Total Words: ${stats.totalWords}`);
    
    console.log('\n📂 حسب النوع / By Type:');
    console.log('─'.repeat(60));
    for (const [type, count] of stats.byType.entries()) {
      console.log(`   ${type}: ${count}`);
    }
    
    console.log('\n🏷️  حسب الفئة / By Category:');
    console.log('─'.repeat(60));
    for (const [category, count] of stats.byCategory.entries()) {
      console.log(`   ${category}: ${count}`);
    }
    console.log('═'.repeat(60));
  }

  /**
   * عرض أمثلة
   * Show examples
   */
  showExamples(): void {
    const categories = [
      FoundationCategory.INITIAL_ENVIRONMENT,
      FoundationCategory.ENTITY_EXISTENCE,
      FoundationCategory.PHYSICAL,
      FoundationCategory.BASIC_ACTIONS
    ];

    console.log('\n📚 أمثلة من القاموس / Examples from Vocabulary');
    console.log('═'.repeat(60));

    for (const category of categories) {
      const words = this.vocab.getWordsByCategory(category);
      if (words.length > 0) {
        console.log(`\n🏷️  ${category}:`);
        const exampleWords = words.slice(0, 3);
        for (const word of exampleWords) {
          console.log(`   • ${word.arabic} (${word.english || 'N/A'}): ${word.coreMeaning}`);
        }
      }
    }
    console.log('═'.repeat(60));
  }

  /**
   * البحث بالمعنى
   * Search by meaning
   */
  searchByMeaning(searchTerm: string): FoundationWord[] {
    const allWords = this.vocab.getAllWords();
    return allWords.filter(word => 
      word.coreMeaning.includes(searchTerm) ||
      word.arabic.includes(searchTerm) ||
      (word.english && word.english.includes(searchTerm))
    );
  }

  /**
   * الحصول على شجرة الكلمات المرتبطة
   * Get related words tree
   */
  getRelatedWordsTree(arabic: string, depth: number = 2): Map<string, FoundationWord[]> {
    const tree = new Map<string, FoundationWord[]>();
    const visited = new Set<string>();
    
    const explore = (word: string, currentDepth: number) => {
      if (currentDepth > depth || visited.has(word)) return;
      
      visited.add(word);
      const related = this.vocab.findRelatedWords(word);
      tree.set(word, related);
      
      if (currentDepth < depth) {
        for (const relatedWord of related) {
          explore(relatedWord.arabic, currentDepth + 1);
        }
      }
    };
    
    explore(arabic, 0);
    return tree;
  }
}

/**
 * Singleton instance
 */
let vocabularyInstance: CompleteFoundationVocabulary | null = null;

/**
 * الحصول على نسخة واحدة من القاموس
 * Get singleton vocabulary instance
 */
export function getCompleteVocabulary(): CompleteFoundationVocabulary {
  if (!vocabularyInstance) {
    vocabularyInstance = new CompleteFoundationVocabulary();
  }
  return vocabularyInstance;
}

/**
 * إعادة تعيين القاموس
 * Reset vocabulary
 */
export function resetVocabulary(): void {
  vocabularyInstance = null;
}

